src/shared/dissect-image.c

```diff
@@ -2004,10 +2004,25 @@ static int mount_partition(
 
         if (p) {
                 if (m->fsmount_fd >= 0) {
+                        bool mount_beneath = FLAGS_SET(flags, DISSECT_IMAGE_MOUNT_BENEATH);
+
                         /* Case #1: Attach existing fsmount fd to the file system */
 
-                        if (move_mount(m->fsmount_fd, "", -EBADF, p, MOVE_MOUNT_F_EMPTY_PATH) < 0)
-                                return -errno;
+                        r = RET_NERRNO(move_mount(m->fsmount_fd,
+                                       "",
+                                       -EBADF,
+                                       p,
+                                       MOVE_MOUNT_F_EMPTY_PATH | (mount_beneath ? MOVE_MOUNT_BENEATH : 0)));
+                        if (mount_beneath && r == -EINVAL) /* Fallback if mount_beneath is not supported */
+                                r = RET_NERRNO(move_mount(m->fsmount_fd,
+                                               "",
+                                               -EBADF,
+                                               p,
+                                               MOVE_MOUNT_F_EMPTY_PATH));
+                        else if (mount_beneath && r >= 0) /* If it is, now remove the old mount */
+                                r = RET_NERRNO(umount2(p, UMOUNT_NOFOLLOW|MNT_DETACH));
+                        if (r < 0)
+                                return r;
 
```

Comment:
what's the result of this code btw, i don't really grok it. you try to atomically replace a mount, but if that doesn't work then you just overmount?

what does MOVE_MOUNT_BENEATH return if we call it on something that actually isn't a mount point? -EINVAL? if so we cannot distinguish the case where the flag is not supported by the kernel from the case where we called it but the target isn't a mount yet, did I get this right?

this definitely needs some comments here, because this is everything but obvious.

Reply:
yeah we get EINVAL both when it's not supported, and when it is but there's no mount to go under. I've updated the comment.